File: CovidWatch/app/src/main/java/com/riskre/covidwatch/data/ContactEventDAO.java
package com.riskre.covidwatch.data;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface ContactEventDAO {

    @Query("SELECT * FROM contact_events")
    List<ContactEvent> getAll();

    @Query("SELECT * FROM contact_events ORDER BY timestamp DESC")
    LiveData<List<ContactEvent>> getAllSortedByDescTimestamp();

    @Query("SELECT * FROM contact_events WHERE identifier = :identifier")
    ContactEvent findByPrimaryKey(String identifier);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(ContactEvent contactEvent);

    @Update
    void update(ContactEvent contactEvent);

    @Query("UPDATE contact_events SET was_potentially_infectious = :wasPotentiallyInfectious WHERE identifier IN (:identifiers)")
    void update(List<String> identifiers, Boolean wasPotentiallyInfectious);

    @Query("DELETE FROM contact_events")
    void deleteAll();
}

File: CovidWatch/app/src/main/java/org/covidwatch/android/data/ContactEventDAO.kt
package org.covidwatch.android.data

import androidx.lifecycle.LiveData
import androidx.paging.DataSource
import androidx.room.*

@Dao
interface ContactEventDAO {
    @get:Query("SELECT * FROM contact_events")
    val all: List<ContactEvent>

    @get:Query("SELECT * FROM contact_events ORDER BY timestamp DESC")
    val allSortedByDescTimestamp: LiveData<List<ContactEvent>>

    @get:Query("SELECT * FROM contact_events ORDER BY timestamp DESC")
    val pagedAllSortedByDescTimestamp: DataSource.Factory<Int, ContactEvent>

    @Query("SELECT * FROM contact_events WHERE identifier = :identifier")
    fun findByPrimaryKey(identifier: String): ContactEvent

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    fun insert(contactEvent: ContactEvent)

    @Update
    fun update(contactEvent: ContactEvent)

    @Update
    fun update(contactEvents: List<ContactEvent>)

    @Query("UPDATE contact_events SET was_potentially_infectious = :wasPotentiallyInfectious WHERE identifier IN (:identifiers)")
    fun update(identifiers: List<String>, wasPotentiallyInfectious: Boolean)

    @Query("UPDATE contact_events SET was_potentially_infectious = '1'")
    fun markAllAsPotentiallyInfectious()

    @Query("UPDATE contact_events SET upload_state = :uploadState WHERE identifier IN (:identifiers)")
    fun update(identifiers: List<String>, uploadState: Int)

    @Query("DELETE FROM contact_events")
    fun deleteAll()
}